src/stan_math_backend/Lower_functions.ml

```diff
@@ -204,141 +189,98 @@ let lower_fun_def (functors : (string, struct_defn) Hashtbl.t)
-      match
-        Hashtbl.find Stan_math_signatures.stan_math_variadic_signatures x
-      with
-      | Some {required_fn_args; _} ->
-          Map.add_multi accum
-            ~key:(Utils.stdlib_distribution_name f)
-            ~data:(List.length required_fn_args)
-      | _ -> Expr.Fixed.Pattern.fold find_functors_expr accum pattern )
-    | _ -> Expr.Fixed.Pattern.fold find_functors_expr accum pattern in
+let get_functor_requirements (p : Program.Numbered.t) =
+  let open Expr.Fixed in
+  let rec find_functors_expr init = function
+    | {pattern= FunApp (StanLib (hof, FnPlain, _), args); _} ->
+        let f accum = function
+          | {pattern= Var name; _} as e
+            when UnsizedType.is_fun_type (Expr.Typed.type_of e) ->
+              Map.add_multi accum
+                ~key:(Utils.stdlib_distribution_name name)
+                ~data:(Lower_expr.functor_type hof)
+          | e -> find_functors_expr accum e in
+        List.fold ~init ~f args
+    | {pattern; _} -> Pattern.fold find_functors_expr init pattern in
   let rec find_functors_stmt accum stmt =
     Stmt.Fixed.(
       Pattern.fold find_functors_expr find_functors_stmt accum stmt.pattern)
   in
   Program.fold find_functors_expr find_functors_stmt String.Map.empty p
 
-(** We need to do a fair bit of bookkeeping to handle the functors necessary for the various
-    higher order functions.
-
-    Each functor needs a forward decl struct before the function,
-    then the function definition,
-    then the actual functor definitions
-  *)
 let collect_functors_functions (p : Program.Numbered.t) : defn list =
-  let (functors : (string, Cpp.struct_defn) Hashtbl.t) =
-    String.Table.create () in
-  let forward_decls = Hash_set.Poly.create () in
-  let reduce_sum_fns = is_fun_used_with_reduce_sum p in
-  let variadic_fns = get_variadic_requirements p in
-  let fun_and_functor_defs =
+  let functor_required = get_functor_requirements p in
+  let fun_has_def =
+    List.filter_map p.functions_block ~f:(function
+      | {fdname; fdbody= Some _; _} -> Some fdname
+      | _ -> None )
+    |> String.Set.of_list in
+  let fun_decls, fun_defns =
     p.functions_block
-    |> List.concat_map
-         ~f:(lower_fun_def functors forward_decls reduce_sum_fns variadic_fns)
-    |> List.map ~f:(fun f -> FunDef f) in
-  let functor_struct_decls =
-    functors |> Hashtbl.data |> List.map ~f:(fun s -> Struct s) in
-  functor_struct_decls @ fun_and_functor_defs
+    |> List.filter_map ~f:(fun d ->
+           if Set.mem fun_has_def d.fdname && Option.is_none d.fdbody then None
+           else
+             let functors =
+               Map.find_multi functor_required d.fdname |> List.stable_dedup
+             in
+             let fn, st = lower_fun_def functors d in
+             let decl, defn = Cpp.split_fun_decl_defn fn in
+             Some
+               (FunDef decl :: List.map ~f:(fun s -> Struct s) st, FunDef defn) )
```

Comment:
I believe the current test failures (`../../ml/stanc3/test/integration/good/overloading/sir_overloaded_odes.hpp:112:8: error: redefinition of ‘struct sir_overloaded_odes_model_namespace::simple_SIR_variadic2_functor__’`) are because of this not properly "collating" functor structs into one struct per name